File: repo/atlas-x402/server/core/verification.ts
export interface PaymentRequirements {
  scheme: string;
  network: string;
  maxAmountRequired: string;
  resource: string;
  description: string;
  mimeType: string;
  payTo: string;
  maxTimeoutSeconds: number;
  asset: string;
  extra?: Record<string, any> | null;
}

export interface PaymentPayload {
  x402Version: number;
  scheme: string;
  network: string;
  payload: Record<string, any>;
}

export interface VerificationResult {
  isValid: boolean;
  invalidReason?: string | null;
}

export interface VerificationOptions {
  paymentPayload: PaymentPayload;
  requirements: PaymentRequirements;
  facilitatorUrl?: string;
  rpcUrl?: string;
}

export async function verifyPayment(options: VerificationOptions): Promise<VerificationResult> {
  const { paymentPayload, requirements, facilitatorUrl, rpcUrl } = options;

  if (facilitatorUrl) {
    return verifyViaFacilitator(paymentPayload, requirements, facilitatorUrl);
  }

  if (requirements.scheme === 'x402+eip712') {
    return verifyEIP712Payment(paymentPayload, requirements, rpcUrl);
  }

  if (requirements.scheme === 'x402+solana') {
    return verifySolanaPayment(paymentPayload, requirements, rpcUrl);
  }

  return {
    isValid: false,
    invalidReason: `Unsupported scheme: ${requirements.scheme}`,
  };
}

async function verifyViaFacilitator(
  paymentPayload: PaymentPayload,
  requirements: PaymentRequirements,
  facilitatorUrl: string
): Promise<VerificationResult> {
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString('base64');

  const response = await fetch(`${facilitatorUrl}/verify`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      x402Version: 1,
      paymentHeader,
      paymentRequirements: requirements,
    }),
  });

  const result = await response.json();
  return {
    isValid: result.isValid || false,
    invalidReason: result.invalidReason || null,
  };
}

async function verifyEIP712Payment(
  paymentPayload: PaymentPayload,
  requirements: PaymentRequirements,
  rpcUrl?: string
): Promise<VerificationResult> {
  const rpc = rpcUrl || (requirements.network === 'base' ? 'https://mainnet.base.org' : 'https://mainnet.infura.io/v3/...');
  const txHash = paymentPayload.payload.transactionHash;

  if (!txHash) {
    return {
      isValid: false,
      invalidReason: 'Missing transaction hash',
    };
  }

  try {
    const response = await fetch(rpc, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'eth_getTransactionReceipt',
        params: [txHash],
        id: 1,
      }),
    });

    const result = await response.json();
    const receipt = result.result;

    if (!receipt || receipt.status !== '0x1') {
      return {
        isValid: false,
        invalidReason: 'Transaction failed or not found',
      };
    }

    if (receipt.to?.toLowerCase() !== requirements.payTo.toLowerCase()) {
      return {
        isValid: false,
        invalidReason: 'Recipient address mismatch',
      };
    }

    return {
      isValid: true,
    };
  } catch (error: any) {
    return {
      isValid: false,
      invalidReason: error.message || 'RPC verification failed',
    };
  }
}

async function verifySolanaPayment(
  paymentPayload: PaymentPayload,
  requirements: PaymentRequirements,
  rpcUrl?: string
): Promise<VerificationResult> {
  const rpc = rpcUrl || 'https://api.mainnet-beta.solana.com';
  const signature = paymentPayload.payload.signature;

  if (!signature) {
    return {
      isValid: false,
      invalidReason: 'Missing transaction signature',
    };
  }

  try {
    const response = await fetch(rpc, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'getTransaction',
        params: [
          signature,
          {
            encoding: 'json',
            maxSupportedTransactionVersion: 0,
          },
        ],
      }),
    });

    const result = await response.json();
    const transaction = result.result;

    if (!transaction || transaction.meta?.err) {
      return {
        isValid: false,
        invalidReason: 'Transaction failed or not found',
      };
    }

    return {
      isValid: true,
    };
  } catch (error: any) {
    return {
      isValid: false,
      invalidReason: error.message || 'RPC verification failed',
    };
  }
}

export { verifyPayment as default };

File: repo/atlas-x402/server/core/verification.py
from typing import Dict, Any, Optional
import aiohttp
import json

async def verify_payment(
    payment_payload: Dict[str, Any],
    requirements: Dict[str, Any],
    facilitator_url: Optional[str] = None,
    rpc_url: Optional[str] = None
) -> Dict[str, Any]:
    if facilitator_url:
        return await verify_via_facilitator(payment_payload, requirements, facilitator_url)
    
    scheme = requirements.get('scheme')
    if scheme == 'x402+eip712':
        return await verify_eip712_payment(payment_payload, requirements, rpc_url)
    elif scheme == 'x402+solana':
        return await verify_solana_payment(payment_payload, requirements, rpc_url)
    
    return {
        'isValid': False,
        'invalidReason': f'Unsupported scheme: {scheme}',
    }

async def verify_via_facilitator(
    payment_payload: Dict[str, Any],
    requirements: Dict[str, Any],
    facilitator_url: str
) -> Dict[str, Any]:
    import base64
    
    payment_header = base64.b64encode(json.dumps(payment_payload).encode()).decode()
    
    async with aiohttp.ClientSession() as session:
        async with session.post(
            f'{facilitator_url}/verify',
            json={
                'x402Version': 1,
                'paymentHeader': payment_header,
                'paymentRequirements': requirements,
            }
        ) as response:
            result = await response.json()
            return {
                'isValid': result.get('isValid', False),
                'invalidReason': result.get('invalidReason'),
            }

async def verify_eip712_payment(
    payment_payload: Dict[str, Any],
    requirements: Dict[str, Any],
    rpc_url: Optional[str]
) -> Dict[str, Any]:
    rpc = rpc_url or ('https://mainnet.base.org' if requirements.get('network') == 'base' else 'https://mainnet.infura.io/v3/...')
    tx_hash = payment_payload.get('payload', {}).get('transactionHash')
    
    if not tx_hash:
        return {
            'isValid': False,
            'invalidReason': 'Missing transaction hash',
        }
    
    async with aiohttp.ClientSession() as session:
        async with session.post(
            rpc,
            json={
                'jsonrpc': '2.0',
                'method': 'eth_getTransactionReceipt',
                'params': [tx_hash],
                'id': 1,
            }
        ) as response:
            result = await response.json()
            receipt = result.get('result')
            
            if not receipt or receipt.get('status') != '0x1':
                return {
                    'isValid': False,
                    'invalidReason': 'Transaction failed or not found',
                }
            
            return {
                'isValid': True,
            }

async def verify_solana_payment(
    payment_payload: Dict[str, Any],
    requirements: Dict[str, Any],
    rpc_url: Optional[str]
) -> Dict[str, Any]:
    rpc = rpc_url or 'https://api.mainnet-beta.solana.com'
    signature = payment_payload.get('payload', {}).get('signature')
    
    if not signature:
        return {
            'isValid': False,
            'invalidReason': 'Missing transaction signature',
        }
    
    async with aiohttp.ClientSession() as session:
        async with session.post(
            rpc,
            json={
                'jsonrpc': '2.0',
                'id': 1,
                'method': 'getTransaction',
                'params': [signature, {'encoding': 'json'}],
            }
        ) as response:
            result = await response.json()
            transaction = result.get('result')
            
            if not transaction or transaction.get('meta', {}).get('err'):
                return {
                    'isValid': False,
                    'invalidReason': 'Transaction failed or not found',
                }
            
            return {
                'isValid': True,
            }





